### ISSUE TITLE
ENOENT in addReader() on epoll reactor when an HTTP client disconnects before receiving a response

### ISSUE THREAD

**Issue by XxSuper** [twisted/twisted#11625]
```
[2022-08-30 11:17:15] [_legacy.py[line:147]] [CRITICAL] [Unhandled Error
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/twisted/python/log.py", line 85, in callWithContext
    return context.call({ILogContext: newCtx}, func, *args, **kw)
  File "/opt/conda/lib/python3.6/site-packages/twisted/python/context.py", line 118, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/opt/conda/lib/python3.6/site-packages/twisted/python/context.py", line 83, in callWithContext
    return func(*args, **kw)
  File "/opt/conda/lib/python3.6/site-packages/twisted/internet/posixbase.py", line 687, in _doReadOrWrite
    why = selectable.doRead()
--- <exception caught here> ---
  File "/opt/conda/lib/python3.6/site-packages/twisted/internet/tcp.py", line 1401, in doRead
    skt, protocol, addr, self, s, self.reactor
  File "/opt/conda/lib/python3.6/site-packages/twisted/internet/tcp.py", line 808, in __init__
    self.startReading()
  File "/opt/conda/lib/python3.6/site-packages/twisted/internet/abstract.py", line 437, in startReading
    self.reactor.addReader(self)
  File "/opt/conda/lib/python3.6/site-packages/twisted/internet/epollreactor.py", line 118, in addReader
    reader, self._reads, self._writes, self._selectables, EPOLLIN, EPOLLOUT
  File "/opt/conda/lib/python3.6/site-packages/twisted/internet/epollreactor.py", line 103, in _add
    self._poller.modify(fd, flags)
builtins.FileNotFoundError: [Errno 2] No such file or directory
```

**Comment by adiroiban** [comment]
Hi, Thanks for the report.

Please add more info about how to reproduce and test this error.

If I am making changes in trying to fix this bug, how can I tell that my changes actually fix the issue?

I see that this is on python 3.6, which is no longer supported.

I am closing this for now as python 3.6 is not longer supported.

Please add more info, see if you can reproduce it on python 3.7 or newer and we can re-open it.

Regards

**Comment by LegendaryM** [comment]
Hi, It happened on python 3.7.
My Twisted version is 22.10.0

```
[2023-03-23 17:19:19,565] [_legacy.py[line:147]] [CRITICAL] [Unhandled Error
Traceback (most recent call last):
  File "/opt/conda/envs/torch1.12/lib/python3.7/site-packages/twisted/python/log.py", line 80, in callWithContext
    return context.call({ILogContext: newCtx}, func, *args, **kw)
  File "/opt/conda/envs/torch1.12/lib/python3.7/site-packages/twisted/python/context.py", line 117, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/opt/conda/envs/torch1.12/lib/python3.7/site-packages/twisted/python/context.py", line 82, in callWithContext
    return func(*args, **kw)
  File "/opt/conda/envs/torch1.12/lib/python3.7/site-packages/twisted/internet/posixbase.py", line 487, in _doReadOrWrite
    why = selectable.doRead()
--- <exception caught here> ---
  File "/opt/conda/envs/torch1.12/lib/python3.7/site-packages/twisted/internet/tcp.py", line 1408, in doRead
    skt, protocol, addr, self, s, self.reactor
  File "/opt/conda/envs/torch1.12/lib/python3.7/site-packages/twisted/internet/tcp.py", line 810, in __init__
    self.startReading()
  File "/opt/conda/envs/torch1.12/lib/python3.7/site-packages/twisted/internet/abstract.py", line 438, in startReading
    self.reactor.addReader(self)
  File "/opt/conda/envs/torch1.12/lib/python3.7/site-packages/twisted/internet/epollreactor.py", line 118, in addReader
    reader, self._reads, self._writes, self._selectables, EPOLLIN, EPOLLOUT
  File "/opt/conda/envs/torch1.12/lib/python3.7/site-packages/twisted/internet/epollreactor.py", line 103, in _add
    self._poller.modify(fd, flags)
builtins.FileNotFoundError: [Errno 2] No such file or directory
```

**Comment by p12tic** [comment]
FYI same backtrace is seen in this bug report to Buildbot: https://github.com/buildbot/buildbot/issues/8306

**Comment by glyph** [comment]
Thanks for the report @p12tic

**Comment by glyph** [comment]
Okay, I think the hint from the buildbot issue, `[-] http client disconnected before results were sent`, gives us enough detail to start investigating.

**Comment by oliver-joos** [comment]
I still see this issue with **Twisted 25.5.0** in latest Buildbot 4.3.0 on Python 3.12.3.

It happened after an (unrelated) crash of Firefox.
In twistd.log I also see `http client disconnected before results were sent`, and then many tracebacks of _poller.modify().

After this issue the Buildbot WebUI stops working properly even in another browser. The newly started Firefox console shows an error that it cannot connect to websocket wss://<URL>/bots/<botname>/ws . I had to restart Buildbot master to recover.

Because the traceback only shows twisted modules and no buildbot modules, I decided to append my findings here.

**Comment by adiroiban** [comment]
Thanks @oliver-joos 
Can you please add your traceback ?

Thanks

**Comment by oliver-joos** [comment]
Of course! Below is a fragment of our Buildbot `twistd.log`.

Often I see multiple `http client disconnected`, then multiple of these Tracebacks, then multiple `Timing out client` in a row.

I would gladly help to solve this issue! Let me know what I should try, or where I can provide further information!

-----
```python
...

2026-08-21 23:08:39+0200 [-] http client disconnected before results were sent
2026-08-21 23:08:40+0200 [-] http client disconnected before results were sent
2026-08-21 23:08:42+0200 [buildbot.www.service.BuildbotSite] Unhandled Error
        Traceback (most recent call last):
          File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/internet/tcp.py", line 1431, in doRead
            transport = self.transport(
          File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/internet/tcp.py", line 834, in __init__
            self.startReading()
          File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/internet/abstract.py", line 456, in startReading
            self.reactor.addReader(self)
          File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/internet/epollreactor.py", line 117, in addReader
            self._add(
          File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/internet/epollreactor.py", line 103, in _add
            self._poller.modify(fd, flags)
        builtins.FileNotFoundError: [Errno 2] No such file or directory

2026-08-21 23:08:43+0200 [buildbot.www.service.BuildbotSite] Unhandled Error
        Traceback (most recent call last):
          File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/internet/tcp.py", line 1431, in doRead
            transport = self.transport(
          File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/internet/tcp.py", line 834, in __init__
            self.startReading()
          File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/internet/abstract.py", line 456, in startReading
            self.reactor.addReader(self)
          File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/internet/epollreactor.py", line 117, in addReader
            self._add(
          File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/internet/epollreactor.py", line 103, in _add
            self._poller.modify(fd, flags)
        builtins.FileNotFoundError: [Errno 2] No such file or directory

2026-08-21 23:09:01+0200 [buildbot.www.service.BuildbotSite] Unhandled Error
        Traceback (most recent call last):
          File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/internet/tcp.py", line 1431, in doRead
            transport = self.transport(
          File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/internet/tcp.py", line 834, in __init__
            self.startReading()
          File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/internet/abstract.py", line 456, in startReading
            self.reactor.addReader(self)
          File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/internet/epollreactor.py", line 117, in addReader
            self._add(
          File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/internet/epollreactor.py", line 103, in _add
            self._poller.modify(fd, flags)
        builtins.FileNotFoundError: [Errno 2] No such file or directory

2026-08-21 23:09:08+0200 [-] Timing out client: IPv4Address(type='TCP', host='127.0.0.1', port=44348)
2026-08-21 23:09:08+0200 [-] Timing out client: IPv4Address(type='TCP', host='127.0.0.1', port=44346)
2026-08-21 23:09:08+0200 [-] Timing out client: IPv4Address(type='TCP', host='127.0.0.1', port=44350)
2026-08-21 23:09:09+0200 [-] Timing out client: IPv4Address(type='TCP', host='127.0.0.1', port=44352)
2026-08-21 23:09:09+0200 [-] Timing out client: IPv4Address(type='TCP', host='127.0.0.1', port=44342)
2026-08-21 23:09:40+0200 [-] Timing out client: IPv4Address(type='TCP', host='127.0.0.1', port=46122)
2026-08-21 23:09:40+0200 [-] Timing out client: IPv4Address(type='TCP', host='127.0.0.1', port=46230)
2026-08-21 23:09:40+0200 [-] Timing out client: IPv4Address(type='TCP', host='127.0.0.1', port=46032)
2026-08-21 23:09:40+0200 [-] Timing out client: IPv4Address(type='TCP', host='127.0.0.1', port=46126)
2026-08-21 23:10:00+0200 [-] Timing out client: IPv4Address(type='TCP', host='127.0.0.1', port=46232)
2026-08-21 23:19:27+0200 [buildbot.www.service.BuildbotSite] Unhandled Error
        Traceback (most recent call last):
          File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/internet/tcp.py", line 1431, in doRead
            transport = self.transport(
          File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/internet/tcp.py", line 834, in __init__

...
```

**Comment by adiroiban** [comment]
Can you try to catch the exception around these lines and with a pdb , try to see how and why doRead is called ?

https://github.com/twisted/twisted/blob/f3155a267fb8847764591c5040455be22c7eb382/src/twisted/internet/tcp.py#L1457-L1459

---

Are there any thread setup here ?

It looks like the TCP port is started... but then during the startup process the TCP port is stopped.

Just a guess

**Comment by oliver-joos** [comment]
Ok, I will try that!
Our Buildbot runs on a headless Cloud server. So I chose to patch `.../internet/tcp.py` to catch tracebacks **and** stack dumps.
Now I have to wait until the issue occurs again.

```
def doRead(self):
    try:
        import traceback
        from datetime import datetime
        ts = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        with open("/tmp/crash.log", "a") as f:
            f.write(f"\n=== {ts} doRead() entry stack ===\n")
            f.writelines(traceback.format_stack())
            f.write(f"=== {ts} end stack ===\n\n")
        ...

    except BaseException:
        ts = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        with open("/tmp/crash.log", "a") as f:
            f.write(f"\n=== {ts} exception in doRead() ===\n")
            traceback.print_exc(file=f)
            f.write(f"=== {ts} end exception ===\n\n")
        ...
```

**Comment by glyph** [comment]
> Our Buildbot runs on a headless Cloud server. So I chose to patch `.../internet/tcp.py` to catch tracebacks **and** stack dumps.

That level of patching is going to create *really* enormous logs, and may slow things down that you'll never see the race condition again. `doRead` is called every single time a single packet arrives.

If you *do* actually manage to reproduce it then we should have a really great diagnostic! But please don't blow up your prod environment just to diagnose this rare issue ;)

**Comment by oliver-joos** [comment]
Thanks for the warning! We are a small team. I can shutdown the server every night for debugging/fixing.

I think, if the stack dumps slow down our Buildbot too much, I could move `traceback.format_stack()` also into `except BaseException`, so that stack dumps are logged only in case of exception.

**Comment by oliver-joos** [comment]
I can reproduce this issue easily with a single browser (Firefox 153.0.4) on the "Console View" page of the Buildbot Web UI by clicking repeatedly on the page refresh button ⇒ refresh - cancel - refresh - cancel ...

I added the proposed patch to `twisted/internet/tcp.py`, and was able to generate this stack dump:

```
File "/home/buildbot/bots/lisa_gateway-env/bin/buildbot", line 6, in <module>
  sys.exit(run())
File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/buildbot/scripts/runner.py", line 868, in run
  sys.exit(subcommandFunction(subconfig))
File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/buildbot/scripts/start.py", line 144, in start
  launchNoDaemon(config)
File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/buildbot/scripts/start.py", line 109, in launchNoDaemon
  twistd.run()
File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/scripts/twistd.py", line 35, in run
  app.run(runApp, ServerOptions)
File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/application/app.py", line 673, in run
  runApp(config)
File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/scripts/twistd.py", line 29, in runApp
  runner.run()
File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/application/app.py", line 374, in run
  self.postApplication()
File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/scripts/_twistd_unix.py", line 268, in postApplication
  self.startReactor(None, self.oldstdout, self.oldstderr)
File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/application/app.py", line 386, in startReactor
  runReactorWithLogging(self.config, oldstdout, oldstderr, self.profiler, reactor)
File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/application/app.py", line 304, in runReactorWithLogging
  reactor.run()
File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/internet/base.py", line 697, in run
  self.mainLoop()
File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/internet/base.py", line 708, in mainLoop
  self.doIteration(t)
File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/internet/epollreactor.py", line 257, in doPoll
  log.callWithLogger(selectable, _drdw, selectable, fd, event)
File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/python/log.py", line 96, in callWithLogger
  return callWithContext({"system": lp}, func, *args, **kw)
File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/python/log.py", line 80, in callWithContext
  return context.call({ILogContext: newCtx}, func, *args, **kw)
File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/python/context.py", line 117, in callWithContext
  return self.currentContext().callWithContext(ctx, func, *args, **kw)
File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/python/context.py", line 82, in callWithContext
  return func(*args, **kw)
File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/internet/posixbase.py", line 491, in _doReadOrWrite
  why = selectable.doRead()
File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/internet/tcp.py", line 1454, in doRead
  f.writelines(traceback.format_stack())
```

**Comment by adiroiban** [comment]
Thanks for tracing this error.
Is this the same error? I see that this fails in doRead() interation, while the previous traceback was in addReader connection startup code.

**Comment by oliver-joos** [comment]
Yes, the above is the **stack dump** in `doRead(): ... except BaseException` just before handling it.

Here is the traceback, immediately after the stack dump above:

```
Traceback (most recent call last):
  File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/internet/tcp.py", line 1431, in doRead
    transport = self.transport(
                ^^^^^^^^^^^^^^^
  File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/internet/tcp.py", line 834, in __init__
    self.startReading()
  File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/internet/abstract.py", line 456, in startReading
    self.reactor.addReader(self)
  File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/internet/epollreactor.py", line 130, in addReader
    self._add(
  File "/home/buildbot/bots/lisa_gateway-env/lib/python3.12/site-packages/twisted/internet/epollreactor.py", line 104, in _add
    self._poller.modify(fd, flags)
FileNotFoundError: [Errno 2] No such file or directory
```
\
⇒ this `FileNotFoundError` is caused by self._poller.modify() on line internet/epollreactor.py:104, which is called by doRead() on line internet/tcp.py:1431 and caught in `except BaseException` on line internet/tcp.py:1454.

In original Twisted 25.5.0 this self._poller.modify() is on line 103. In my Twisted 25.5.0 I inserted a `try...except` around it. That's why it is on line 104.

**Comment by glyph** [comment]
@oliver-joos so, it seems clear that what's happening is an out-of-sequence `socket.close()` occurring before the socket can be fully removed from `epoll`.  This *ought* to be an invariant that Twisted enforces, but our methods for doing this are not particularly rigorous.

Here's some narration of what's going on.

The traceback is happening in `Port.doRead` calling `Server.__init__`, which means it is being called immediately after `accept()` has returned something.

The `modify` that is failing is the addition of the new inbound connection's file descriptor, which means that we should be looking at places where a file descriptor would still be tracked by the reactor, but closed somewhere else, because this new accepted socket is stomping on some exiting connection that has been forcibly closed.

With all that said, this looks very suspicious to me:

https://github.com/buildbot/buildbot/blob/61f2f4118098dbbd22cf9212e2b470cea7c505ba/master/buildbot/test/util/site.py#L44-L46

`transport.connectionLost` really shouldn't be a public API, and it's only really safe to call it from `twisted.internet.posixbase._DisconnectSelectableMixin._disconnectSelectable`, which is what tries to ensure that `removeReader` and `removeWriter` are called *before* `readConnectionLost` or `connectionLost` .

I'm not ready to call this the smoking gun, because it seems to only be called from unit tests, but it definitely looks like a potential problem if it is ever invoked at runtime in production code somehow.

**Comment by glyph** [comment]
*Specifically*, this means that the brand-new socket we just accepted and we are trying to add to `self._reads` is already in `self._writes`, and attempting to modify it gives us an I/O error.

Experimenting, with this, it seems that epoll will give this type of error if you register, then modify, then close a file descriptor, until you re-register it, even if it *is* open at the time of the modification:

```
Python 3.14.0 (main, Nov 18 2025, 07:48:54) [GCC 14.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from select import epoll
>>> from socket import socket
>>> s = socket()
>>> e = epoll()
>>> e.add(s.fileno())
Traceback (most recent call last):
  File "<python-input-4>", line 1, in <module>
    e.add(s.fileno())
    ^^^^^
AttributeError: 'select.epoll' object has no attribute 'add'
>>> e.register(s.fileno())
>>> e.modify(s.fileno(), EPOLLIN)
Traceback (most recent call last):
  File "<python-input-6>", line 1, in <module>
    e.modify(s.fileno(), EPOLLIN)
                         ^^^^^^^
NameError: name 'EPOLLIN' is not defined
>>> from select import epoll, EPOLLIN, EPOLLOUT
>>> e.modify(s.fileno(), EPOLLOUT)
>>> s.fileno()
5
>>> s.close()
>>> s2 = socket()
>>> s2.fileno()
5
>>> e.modify(s.fileno(), EPOLLOUT)
Traceback (most recent call last):
  File "<python-input-13>", line 1, in <module>
    e.modify(s.fileno(), EPOLLOUT)
    ~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^
ValueError: file descriptor cannot be a negative integer (-1)
>>> e.modify(s2.fileno(), EPOLLOUT)
Traceback (most recent call last):
  File "<python-input-14>", line 1, in <module>
    e.modify(s2.fileno(), EPOLLOUT)
    ~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory
>>> e.modify(s2.fileno(), EPOLLOUT)
Traceback (most recent call last):
  File "<python-input-15>", line 1, in <module>
    e.modify(s2.fileno(), EPOLLOUT)
    ~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory
>>> e.register(s2.fileno(), EPOLLOUT)
>>> e.modify(s2.fileno(), EPOLLOUT)
>>> 
```

which implies that we CAN actually detect closed-from-underneath-us fairly reliably via epoll (more reliably than any of the other reactors, even!), if this behavior can be relied upon.

**Comment by glyph** [comment]
I think a good next step in an investigation would be to instrument this exception code path to note which object is being prematurely closed, and what state it is in (i.e.: has it had `connectionLost` invoked on it?) and then we might have a clearer picture of where the bug is.

**Comment by glyph** [comment]
I don't think we could enable `twisted.internet.test.test_fdset.ReactorFDSetTestsBuilder.test_lostFileDescriptor` directly, because that relies too much on the precise iteration of the loop *detecting* the lost file descriptor (which epoll indeed cannot do because that file descriptor is no longer *in* the set, and thus will never have an event detected on it), but we could definitely write an Epoll-specific version which adopts a new socket under the same file descriptor and make sure that it can still be added and the old one will be cycled out.

This does still indicate a potentially dangerous situation (something might be randomly closing file descriptors out from under Twisted, corrupting its state) and we should still be logging an exception, but we should be able to at least accept the new connection and partially clear out the broken state, maybe invoking `.connectionLost()` on the way out for the now-closed thing.

**Comment by oliver-joos** [comment]
Unfortunately I am not familiar with Twisted. I work more on MicroPython projects like a pure-Python socket library for WiFi chips with AT interface. So I do know about closing/reusing connections, socket error handling etc.

You showed how `e.modify()` may raise `FileNotFoundError` without Twisted or Buildbot - that's great!
I reproduced that:

```
Python 3.12.3 (main, Jun 19 2026, 12:46:00) [GCC 13.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from select import epoll, EPOLLIN, EPOLLOUT
>>> from socket import socket
>>> s = socket()
>>> s.fileno()
3
>>> e = epoll()
>>> e.register(s.fileno())
>>> s.close()
>>> 
>>> s2 = socket()
>>> s2.fileno()
3
>>> e.modify(s2.fileno(), EPOLLOUT)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
FileNotFoundError: [Errno 2] No such file or directory
```

I think to prevent the exception we must prevent the guilty `s.close()` in Twisted, right?
To find when/where/why the socket is closed I could try to monkey-patch `socket.close` to dump the stack and other system states. Possible? Helpful? Suggestions what to dump?

**Comment by glyph** [comment]
> I think to prevent the exception we must prevent the guilty `s.close()` in Twisted, right? To find when/where/why the socket is closed I could try to monkey-patch `socket.close` to dump the stack and other system states. Possible? Helpful? Suggestions what to dump?

The thing is the file descriptor *could* be closed by anybody; Twisted, buildbot, some 3rd party library… it could be closed at the Twisted level, the Python socket level, or it could be closed at the `os._close` level, so tracing this is tricky, especially because so many file descriptors are validly opened and closed.

One additional detail here that is useful to think about though: we are inside the `if fd not in primary` block of the call to `addReader`, which means that we are in the specific case where not only is the other file descriptor registered, but it's *only registered as a **write** descriptor*, which suggests that it is in a half-closed state.  And nothing in `twisted.web` seems to implement `IHalfCloseableProtocol`, which would make it challenging to put an HTTP connection into that state.

**Comment by glyph** [comment]
The most likely culprit, I think, to be a write-only file descriptor, would be a `ProcessWriter`. So it might be worth investigating buildbot's use of `spawnProcess` to see if anything odd is going on there.

**Comment by glyph** [comment]
okay github code search is broken, this is definitely not right:

https://github.com/search?q=repo%3Abuildbot%2Fbuildbot+spawnProcess&type=code

https://github.com/buildbot/buildbot/blob/61f2f4118098dbbd22cf9212e2b470cea7c505ba/worker/buildbot_worker/runprocess.py#L675C49-L682

**Comment by glyph** [comment]
@oliver-joos the branch in #12826 now contains some logging instrumentation which could tell us *which* kind of file descriptor is getting replaced, and allow things to continue without a traceback. do you think you could try deploying that to see what it says the old/bad FD is? If this confirms my hypothesis that it's a subprocess pipe, that would tell us something interesting. If it's an HTTP connection then at least that's some new information :)

**Comment by glyph** [comment]
I guess I touched the hot potato

**Comment by oliver-joos** [comment]
Thanks for helping with the potato!

Below is my first result. Looks like it's a HTTP connection. I already saw this `AttributeError` without patched `epollreactor.py`. This could be just another victim of the closing fd. Or is it a hint to the cause of it all?

Tell me if you need more/other traceback logs!

```
...
2026-09-01 15:48:23+0200 [-] http client disconnected before results were sent
2026-09-01 15:48:24+0200 [-] http client disconnected before results were sent
2026-09-01 15:48:24+0200 [-] http client disconnected before results were sent
2026-09-01 15:48:25+0200 [-] http client disconnected before results were sent
2026-09-01 15:48:28+0200 [-] http client disconnected before results were sent
2026-09-01 15:48:28+0200 [-] while handling API request
        Traceback (most recent call last):
          File "/home/buildbot/bots/KNX_Vision-env/lib/python3.12/site-packages/buildbot/www/rest.py", line 165, in handleErrors
            yield
          File "/home/buildbot/bots/KNX_Vision-env/lib/python3.12/site-packages/buildbot/www/rest.py", line 419, in renderRest
            yield threads.deferToThread(V2RootResource._write_json_data, request, encoder, data)
          File "/home/buildbot/bots/KNX_Vision-env/lib/python3.12/site-packages/twisted/python/threadpool.py", line 269, in inContext
            result = inContext.theWork()  # type: ignore[attr-defined]
          File "/home/buildbot/bots/KNX_Vision-env/lib/python3.12/site-packages/twisted/python/threadpool.py", line 285, in <lambda>
            inContext.theWork = lambda: context.call(  # type: ignore[attr-defined]
          File "/home/buildbot/bots/KNX_Vision-env/lib/python3.12/site-packages/twisted/python/context.py", line 117, in callWithContext
            return self.currentContext().callWithContext(ctx, func, *args, **kw)
          File "/home/buildbot/bots/KNX_Vision-env/lib/python3.12/site-packages/twisted/python/context.py", line 82, in callWithContext
            return func(*args, **kw)
          File "/home/buildbot/bots/KNX_Vision-env/lib/python3.12/site-packages/buildbot/www/rest.py", line 513, in _write_json_data
            request.write(unicode2bytes(chunk))
          File "/home/buildbot/bots/KNX_Vision-env/lib/python3.12/site-packages/twisted/web/server.py", line 267, in write
            _HTTPRequest.write(self, data)
          File "/home/buildbot/bots/KNX_Vision-env/lib/python3.12/site-packages/twisted/web/http.py", line 1327, in write
            self.channel.writeSequence(toChunk(data))
        builtins.AttributeError: 'NoneType' object has no attribute 'writeSequence'

2026-09-01 15:48:28+0200 [-] http client disconnected before results were sent
2026-09-01 15:48:32+0200 [-] http client disconnected before results were sent
2026-09-01 15:48:35+0200 [-] invalid closed fd (23) in epoll for <_GenericHTTPChannelProtocol #229 on 8015> being replaced by <_GenericHTTPChannelProtocol #292 on 8015>
2026-09-01 15:48:35+0200 [-] while handling API request
        Traceback (most recent call last):
          File "/home/buildbot/bots/KNX_Vision-env/lib/python3.12/site-packages/buildbot/www/rest.py", line 165, in handleErrors
            yield
          File "/home/buildbot/bots/KNX_Vision-env/lib/python3.12/site-packages/buildbot/www/rest.py", line 419, in renderRest
            yield threads.deferToThread(V2RootResource._write_json_data, request, encoder, data)
          File "/home/buildbot/bots/KNX_Vision-env/lib/python3.12/site-packages/twisted/python/threadpool.py", line 269, in inContext
            result = inContext.theWork()  # type: ignore[attr-defined]
          File "/home/buildbot/bots/KNX_Vision-env/lib/python3.12/site-packages/twisted/python/threadpool.py", line 285, in <lambda>
            inContext.theWork = lambda: context.call(  # type: ignore[attr-defined]
          File "/home/buildbot/bots/KNX_Vision-env/lib/python3.12/site-packages/twisted/python/context.py", line 117, in callWithContext
            return self.currentContext().callWithContext(ctx, func, *args, **kw)
          File "/home/buildbot/bots/KNX_Vision-env/lib/python3.12/site-packages/twisted/python/context.py", line 82, in callWithContext
            return func(*args, **kw)
          File "/home/buildbot/bots/KNX_Vision-env/lib/python3.12/site-packages/buildbot/www/rest.py", line 513, in _write_json_data
            request.write(unicode2bytes(chunk))
          File "/home/buildbot/bots/KNX_Vision-env/lib/python3.12/site-packages/twisted/web/server.py", line 267, in write
            _HTTPRequest.write(self, data)
          File "/home/buildbot/bots/KNX_Vision-env/lib/python3.12/site-packages/twisted/web/http.py", line 1327, in write
            self.channel.writeSequence(toChunk(data))
        builtins.AttributeError: 'NoneType' object has no attribute 'writeSequence'

2026-09-01 15:48:35+0200 [-] http client disconnected before results were sent
2026-09-01 15:48:38+0200 [-] invalid closed fd (22) in epoll for <_GenericHTTPChannelProtocol #291 on 8015> being replaced by <_GenericHTTPChannelProtocol #320 on 8015>
2026-09-01 15:48:39+0200 [-] http client disconnected before results were sent
2026-09-01 15:48:39+0200 [-] http client disconnected before results were sent
2026-09-01 15:48:40+0200 [-] http client disconnected before results were sent
2026-09-01 15:48:41+0200 [-] http client disconnected before results were sent
```

**Comment by glyph** [comment]
AHA!

This is the bug:

```
          File "/home/buildbot/bots/KNX_Vision-env/lib/python3.12/site-packages/buildbot/www/rest.py", line 419, in renderRest
            yield threads.deferToThread(V2RootResource._write_json_data, request, encoder, data)
```

Buildbot is calling `request.write` from a thread; that is not thread safe, and is thus corrupting the reactor's state. This is not a bug in Twisted.

**Comment by glyph** [comment]
I think it's still useful to incorporate this change to make this type of error more debuggable and logs less noisy in the future, but hopefully Buildbot can finally fix this on their end, and now we know why we're not seeing it in other applications :)

**Comment by oliver-joos** [comment]
Thank you very much @glyph for your detailed analysis!

I understand that calling requests.write() in a thread can cause glitches like this AttributeError. But are you sure that this also explains "invalid closed fd (22) in epoll"? This occurs 3 or even 7 seconds after the AttributeError.

Anyway, I have to look into it more closely, and then will open or extend an issue for Buildbot.

And I agree that Twisted could emit nicer log messages in such a case. Even a hint to multithreading without proper locking could help.

**Comment by glyph** [comment]
> I understand that calling requests.write() in a thread can cause glitches like this AttributeError. But are you sure that this also explains "invalid closed fd (22) in epoll"? This occurs 3 or even 7 seconds after the AttributeError.

I am not 100% sure I can trace through the exact place where this happens, but I am almost sure. It's difficult to do that kind of tracing in this situation where there's a thread-safety violation, because *any* line of code can go out of sequence with any other line of code.

However, the logic goes something like this: `.write` (which is happening in a thread, in parallel with the reactor itself) calls `.addReader`, which registers the selectable as present in `selectables`. If this happens *while* the connection is closing, then you can get a situation where the main reactor thread is setting up `_disconnectSelectable` to call `socket.close()` in the next iteration around the loop, at the same moment that we are putting the socket's fileno back into the set. When the socket closes, we clean up the transport, calling `loseConnection` on the main thread, which is why `self.channel` is `None`. But, that `write` is still proceeding on the thread, which is why its channel disappears _out from under it_ as it goes, and also why the FD is now in the wrong place.

Basically the transport is torn down normally, but *as* it's being torn down, we are simultaneously screwing up the bookkeeping state around it, which causes it to look as though the reactor has lost track of that state.

**Comment by oliver-joos** [comment]
I can’t get that deeply into the details of Twisted.
What I see in `buildbot/www/rest.py`: the `AttributeError` is raised in `handleErrors()`. That looks familiar to me: trying to write the peer about a connection error - useless!
I propose to continue this topic under buildbot/buildbot/issues/5107 .
Then this issue can focus on improving error logging of Twisted in case of reused file descriptors.

**Comment by glyph** [comment]
@oliver-joos if you can use a version of buildbot that has https://github.com/buildbot/buildbot/pull/9143 incorporated into it, do you still see this issue? If you can use that as the same time as Twisted, we can see if anything *else* triggers this

**Comment by oliver-joos** [comment]
**UPDATE:** We use the patched `buildbot/www/rest.py` since 2 weeks for multiple Buildbots in a production environment,
and we have no more FileNotFoundError or AttributeError in any twistd.log!

Many thanks again to @glyph!

As mentioned above, Twisted did not cause this issue, but could improve its error logging in case of reused file descriptors.
To see if or when this fix is merged into Buildbot, see https://github.com/buildbot/buildbot/issues/5107

**Comment by bartvdbraak** [comment]
@oliver-joos Thanks, this was bugging me too for our Buildbot instances.
